Specifications/QIR/Callables.md

```diff
@@ -205,9 +217,80 @@ The following snippet of LLVM code could be generated:
+  Applying this function to a `%Callable` whose implementation table has a null
+  "adj" entry should cause a runtime failure.
+- For `__quantum__rt__callable_make_control`, the count of the number of 
+  times the function has been applied must be tracked. The `Controlled` functor
+  is not its own inverse.
+  Applying this function to a `%Callable` whose implementation table has a null
+  "ctl" entry should cause a runtime failure.
+- The order of applying these two functions does not need to be tracked. The
+  `Adjoint` and `Controlled` functors commute.
+
+When `__quantum__rt__callable_invoke` is called, the entry in the callable's
+implementation table to be used is selected as follows:
+
+- If the adjoint parity is even and the controlled count is zero, the "body"
+  entry (index 0) should be used.
+- If the adjoint parity is odd and the controlled count is zero, the "adj"
+  entry (index 1) should be used.
+- If the adjoint parity is even and the controlled count is greater than zero,
+  the "ctl" entry (index 2) should be used.
+- If the adjoint parity is odd and the controlled count is greater than zero,
+  the "ctladj" entry (index 3) should be used.
+
+If the controlled count is greater than one, then `__quantum__rt__callable_invoke`
+also needs to do some manipulation of the input tuple.
+Each application of the `Controlled` functor modifies the signature of the
+specialization by adding replacing the current argument tuple with a two-tuple
+containing the array of control qubits as the first element and a tuple of
+the remaining arguments as the second tuple.
+
+For instance, if the base callable expects an argument tuple
+`{ %TupleHeader, i64, %Qubit* }`, then the `Controlled` version expects
+`{ %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }`, and the twice-`Controlled`
+version expects
+`{ %TupleHeader, %Array*, { %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }* }`.
+The "ctl" implementation function always expects
+`{ %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }`.
+Thus, if the controlled count is greater than 1,
+`__quantum__rt__callable_invoke` needs to disassemble the argument tuple,
+concatenate the control qubit arrays, and form the expected argument tuple.
+
+One additional complexity is that the above is modified slightly if the base
+callable expects an argument tuple with exactly one element.
+In this case, the `Controlled` version expects a two-element tuples as
+above, but with the actual base argument as the second element.
+For instance, if the base callable expects `{ %TupleHeader, %Qubit* }`,
+the singly-`Controlled` version expects `{ %TupleHeader, %Array*, %Qubit* }`
+rather than `{ %TupleHeader, %Array*, { %TupleHeader, %Qubit* }* }`.
+This means that the second element of the singly-`Controlled` argument tuple
+is not always a pointer to a struct, and in particular may have variable length
+up to the size of a `%Range`.
```

Comment:
Is it clear how things should work if the base argument type is Unit (admittedly a rather horrible edge case that likely is not of real use)?